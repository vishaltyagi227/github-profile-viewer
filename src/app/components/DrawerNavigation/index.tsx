import React, { useState } from 'react';
import { NavLink } from 'react-router-dom';
import { Box, Drawer, List, ListItem, IconButton } from '@mui/material';
import MenuIcon from '@mui/icons-material/Menu';

import { ROUTE, PROTECTED_ROUTE, ORIGIN } from '@constant/index';
import INavbar from '@containers/Navbar/type';
import Link from '@components/Link';
import Button from '@components/Button';
import useStyles from './style';

const DrawerNavigation = ({ isLogged, handleClick }: INavbar) => {
  const classes = useStyles();
  const [open, setOpen] = useState(false);

  const handleDrawerOpen = () => {
    setOpen(true);
  };

  const handleDrawerClose = () => {
    setOpen(false);
  };

  return (
    <>
      <IconButton className={classes.hamburger} onClick={handleDrawerOpen}>
        <MenuIcon />
      </IconButton>
      <Drawer anchor={ORIGIN.right} open={open} onClose={handleDrawerClose}>
        <>
          <Box onClick={handleDrawerClose} className={classes.wrapper}>
            <List>
              <ListItem className={classes.navItem}>
                <Link
                  btn
                  component={NavLink}
                  to={ROUTE.Home}
                  activeClassName={classes.active}
                  className={classes.navLink}
                  exact
                >
                  Search
                </Link>
              </ListItem>
              {isLogged && (
                <>
                  <ListItem className={classes.navItem}>
                    <Link
                      btn
                      component={NavLink}
                      to={PROTECTED_ROUTE.Follow}
                      activeClassName={classes.active}
                      className={classes.navLink}
                      exact
                    >
                      who to Follow
                    </Link>
                  </ListItem>
                  <ListItem className={classes.navItem}>
                    <Link
                      btn
                      component={NavLink}
                      to={PROTECTED_ROUTE.Profile}
                      activeClassName={classes.active}
                      className={classes.navLink}
                      exact
                    >
                      Profile
                    </Link>
                  </ListItem>
                </>
              )}
              {isLogged ? (
                <>
                  <ListItem className={classes.navItem}>
                    <Button onClick={handleClick} className={classes.navLink}>
                      Logout
                    </Button>
                  </ListItem>
                </>
              ) : (
                <>
                  <ListItem className={classes.navItem}>
                    <Link
                      btn
                      component={NavLink}
                      to={ROUTE.Login}
                      activeClassName={classes.active}
                      className={classes.navLink}
                      exact
                    >
                      Login
                    </Link>
                  </ListItem>
                </>
              )}
            </List>
          </Box>
        </>
      </Drawer>
    </>
  );
};

export default DrawerNavigation;
