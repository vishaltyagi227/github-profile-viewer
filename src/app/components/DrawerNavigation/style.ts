import { Theme } from '@mui/material/styles';
import { createStyles, makeStyles } from '@mui/styles';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    wrapper: {
      minWidth: theme.pxToRem(250),
      height: '100vh',
      paddingTop: theme.spacing(2),
      backgroundColor: theme.customPallete.primary.main,
      color: theme.customPallete.secondary.main,
    },
    navItem: {
      marginTop: theme.spacing(1),
    },
    navLink: {
      color: 'inherit',
      '&:hover': {
        backgroundColor: theme.customPallete.primary.light,
      },
    },
    active: {
      backgroundColor: theme.customPallete.primary.light,
    },
    hamburger: {
      color: 'inherit',
    },
  })
);

export default useStyles;
