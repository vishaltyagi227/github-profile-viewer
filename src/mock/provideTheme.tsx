import React, { ReactElement } from 'react';
import { ThemeProvider } from '@mui/material/styles';
import theme from '@theme/index';

export default (ui: ReactElement): ReactElement => (
  <ThemeProvider theme={theme}>{ui}</ThemeProvider>
);
