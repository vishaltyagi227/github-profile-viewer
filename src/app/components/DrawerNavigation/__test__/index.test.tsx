import React from 'react';
import {
  render,
  RenderResult,
  fireEvent,
  cleanup,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import provideTheme from '@mock/provideTheme';
import { BrowserRouter } from 'react-router-dom';
import { ROUTE, PROTECTED_ROUTE } from '@constant/index';
import DrawerNavigation from '../index';

describe('<DrawerNavigation />when user not Logged', () => {
  let container: RenderResult;
  beforeEach(() => {
    container = render(
      provideTheme(
        <BrowserRouter>
          <DrawerNavigation isLogged={false} handleClick={jest.fn()} />
        </BrowserRouter>
      )
    );
  });

  afterEach(cleanup);

  it('Should render hamburger menu', () => {
    expect(container.getByRole('button')).toBeInTheDocument();
  });

  it('Should open menu when user click on the hamburger and have link with proper route', () => {
    fireEvent.click(container.getByRole('button'));

    expect(container.getByRole('link', { name: /Search/i })).toHaveAttribute(
      'href',
      ROUTE.Home
    );

    expect(container.getByRole('link', { name: /login/i })).toHaveAttribute(
      'href',
      ROUTE.Login
    );

    expect(
      container.queryByRole('link', { name: /who to follow/i })
    ).not.toBeInTheDocument();

    expect(
      container.queryByRole('link', { name: /profile/i })
    ).not.toBeInTheDocument();

    expect(
      container.queryByRole('button', { name: /logout/i })
    ).not.toBeInTheDocument();
  });
});

describe('<DrawerNavigation /> when user Logged', () => {
  let container: RenderResult;
  beforeEach(() => {
    container = render(
      provideTheme(
        <BrowserRouter>
          <DrawerNavigation isLogged handleClick={jest.fn()} />
        </BrowserRouter>
      )
    );
  });

  afterEach(cleanup);

  it('Should render hamburger menu', () => {
    expect(container.getByRole('button')).toBeInTheDocument();
  });

  it('Should open menu when user click on the hamburger and have link with proper route', () => {
    fireEvent.click(container.getByRole('button'));

    expect(container.getByRole('link', { name: /Search/i })).toHaveAttribute(
      'href',
      ROUTE.Home
    );

    expect(
      container.queryByRole('link', { name: /login/i })
    ).not.toBeInTheDocument();

    expect(
      container.getByRole('link', { name: /who to follow/i })
    ).toHaveAttribute('href', PROTECTED_ROUTE.Follow);

    expect(container.getByRole('link', { name: /profile/i })).toHaveAttribute(
      'href',
      PROTECTED_ROUTE.Profile
    );

    expect(
      container.getByRole('button', { name: /logout/i })
    ).toBeInTheDocument();
  });
});

describe('<DrawerNavigation /> Snapshot', () => {
  afterEach(cleanup);

  it('Should match with snapshot when user not login', () => {
    const tree = renderer
      .create(
        provideTheme(
          <BrowserRouter>
            <DrawerNavigation isLogged={false} handleClick={jest.fn} />
          </BrowserRouter>
        )
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should match with snapshot when user logged', () => {
    const tree = renderer
      .create(
        provideTheme(
          <BrowserRouter>
            <DrawerNavigation isLogged handleClick={jest.fn} />
          </BrowserRouter>
        )
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
