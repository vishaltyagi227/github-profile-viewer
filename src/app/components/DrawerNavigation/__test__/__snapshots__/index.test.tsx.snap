// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DrawerNavigation /> Snapshot Should match with snapshot when user logged 1`] = `
<button
  className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium makeStyles-hamburger-26 css-78trlr-MuiButtonBase-root-MuiIconButton-root"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
    data-testid="MenuIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
    />
  </svg>
  <span
    className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`<DrawerNavigation /> Snapshot Should match with snapshot when user not login 1`] = `
<button
  className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium makeStyles-hamburger-26 css-78trlr-MuiButtonBase-root-MuiIconButton-root"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
    data-testid="MenuIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
    />
  </svg>
</button>
`;
